### PR TITLE
Write and test new Pytorch dataloaders

### PR DESCRIPTION
**Pull request by ani0203-97**
- Write new PyTorch data loader
- Write script to test data loader behavior without having to run train step

### CODE REVIEW

File: nnunetv2/training/dataloading/nnunet_dataset.py [hunk 1]
```diff
@@ -110,6 +112,219 @@ def load_case(self, key):
 
         return data, seg, entry['properties']
 
+class nnUNetPytorchDataset(Dataset, nnUNetDataset):
```
Comment (gnedivad) [inline nnunetv2/training/dataloading/nnunet_dataset.py]: Break it out into a separate file `nnunetv2/training/dataloading/pytorch_nnunet_dataset.py`

Comment (gnedivad) [reply nnunetv2/training/dataloading/nnunet_dataset.py]: Recommend against inheritance if we eventually want to remove depedency on `nnUNetDataset`.

File: nnunetv2/training/dataloading/nnunet_dataset.py [hunk 2]
```diff
@@ -110,6 +112,220 @@ def load_case(self, key):
...
+            seg = np.load(entry['data_file'])['seg']
+
+        if 'seg_from_prev_stage_file' in entry.keys():
+            if isfile(entry['seg_from_prev_stage_file'][:-4] + ".npy"):
+                seg_prev = np.load(entry['seg_from_prev_stage_file'][:-4] + ".npy", 'r')
+            else:
+                seg_prev = np.load(entry['seg_from_prev_stage_file'])['seg']
+            seg = np.vstack((seg, seg_prev[None]))
+
+        return data, seg, entry['properties']
```
Comment (gnedivad) [inline nnunetv2/training/dataloading/nnunet_dataset.py]: In the existing implementation should be something like
```python
def load_case(self, index: int) -> Tuple[np.ndarray, np.ndarray, PropertiesDict]:
    key = list(self.keys())[idx]
    return super().load_case(key)
```

With the updated way:
```python
def load_case(self, index: int) -> Tuple[np.ndarray, np.ndarray, PropertiesDict]:
    key = list(self.source_nnunet_dataset.keys())[idx]
    return self.source_nnunet_dataset.load_case(key)
```

File: nnunetv2/training/dataloading/nnunet_dataset.py [hunk 3]
```diff
@@ -110,6 +112,220 @@ def load_case(self, key):
...
+        - patch_size
+        - has_ignore
+
+        Functions from the nnUNet Dataloadert that need to be moved to this dataset class. 
+        - get_bbox
+        - need to pad (which is) - self.need_to_pad = (np.array(patch_size) - np.array(final_patch_size)).astype(int)
+
+        '''                
+        
+        # Read in ENTIRE CT and Segmentation from Disk and the properties
```
Comment (gnedivad) [inline nnunetv2/training/dataloading/nnunet_dataset.py]: Same comment about reusing existing code as above.

File: nnunetv2/training/dataloading/nnunet_dataset.py [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
-import shutil
+from batchgenerators.utilities.file_and_folder_operations import (isfile, join,
+                                                                  load_pickle)
+from torch.utils.data import Dataset
 
-from batchgenerators.utilities.file_and_folder_operations import join, load_pickle, isfile
 from nnunetv2.training.dataloading.utils import get_case_identifiers
+from nnunetv2.utilities.label_handling.label_handling import LabelManager
+
+import os
```
Comment (abhijeetshenoi) [inline nnunetv2/training/dataloading/nnunet_dataset.py]: This import is a duplicate

File: nnunetv2/training/dataloading/pytorch_nnunet_dataset.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(self, folder: str, patch_size: Union[List[int], Tuple[int, ...], np.ndarray],
+                 final_patch_size: Union[List[int], Tuple[int, ...], np.ndarray],
+                 label_manager: LabelManager,
+                 transform,                 
+                 case_identifiers: List[str] = None,                 
+                 oversample_foreground_percent: float = 0.0,
+                 num_images_properties_loading_threshold: int = 2000,
+                 folder_with_segs_from_previous_stage: str = None):
+        # Initialize 
+        super().__init__()
```
Comment (abhijeetshenoi) [inline nnunetv2/training/dataloading/pytorch_nnunet_dataset.py]: I don't think you need this for a pytorch dataset

File: nnunetv2/training/dataloading/pytorch_nnunet_dataset.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        for c in case_identifiers:
+            self.dataset[c] = {}
+            self.dataset[c]['data_file'] = os.path.join(folder, f"{c}.npz")
+            self.dataset[c]['properties_file'] = os.path.join(folder, f"{c}.pkl")
+            if folder_with_segs_from_previous_stage is not None:
+                self.dataset[c]['seg_from_prev_stage_file'] = os.path.join(folder_with_segs_from_previous_stage, f"{c}.npz")
+
+        if len(case_identifiers) <= num_images_properties_loading_threshold:
+            for i in self.dataset.keys():
+                self.dataset[i]['properties'] = load_pickle(self.dataset[i]['properties_file'])
```
Comment (abhijeetshenoi) [inline nnunetv2/training/dataloading/pytorch_nnunet_dataset.py]: Move to separate function.
Also, can you explain what this is doing?
- What is "previous_stage" for previous stage files and seg
- What is num_images_properties_loading_threshold
- Why are we saving references to properties files and also loading some different ones?

File: nnunetv2/training/dataloading/test_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                        help='[OPTIONAL] Use this flag to specify a custom trainer. Default: nnUNetTrainer. To run the Pytorch DataLoader, use "nnUNetTrainerPyTorchDataloader"')
+    parser.add_argument('-p', type=str, required=False, default='nnUNetPlans',
+                        help='[OPTIONAL] Use this flag to specify a custom plans identifier. Default: nnUNetPlans')
+    parser.add_argument('-num_gpus', type=int, default=1, required=False,
+                        help='Specify the number of GPUs to use for training')
+    parser.add_argument("--use_compressed", default=False, action="store_true", required=False,
+                        help="[OPTIONAL] If you set this flag the training cases will not be decompressed. Reading compressed "
+                             "data is much more CPU and (potentially) RAM intensive and should only be used if you "
+                             "know what you are doing")
+    parser.add_argument('-device', type=str, default='cuda', required=False,
```
Comment (abhijeetshenoi) [inline nnunetv2/training/dataloading/test_dataloader.py]: It's more typical to use -- for options and - for flags

File: nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py
Comment (abhijeetshenoi) [inline nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py:1]: These changes need to be removed, correct?

File: nnunetv2/training/dataloading/pytorch_nnunet_dataset.py (hunk 6 above)
Comment (ani0203-97) [reply nnunetv2/training/dataloading/pytorch_nnunet_dataset.py]: - nUNet has a functionality to train cascade models, where you have a first stage model which is low resolution and a second stage model that takes in input CT and the low resolution segmentation mask and predicts high resolution segmentation mask. The "seg_from_prev_stage_file" is to facilitate that. We don't need it as of now and it can be removed. 
- num_images_properties_loading_threshold - This is a variable they use to decide whether to load the input image properties into RAM or read them on the fly from the pickle file. The properties contain info such as where the foreground pixels are, the orginal size of the image and other meta data. The default value of this is kept at 0, which means nnUNet always reads metadata (most useful of which is foreground pixel location) on the fly. I don't see the value in having this either and am ok making a choice about reading metadata on the fly or loading into RAM and keeping that fixed. 
- I don't understand the last q

File: nnunetv2/training/dataloading/test_dataloader.py (hunk 7 above)
Comment (ani0203-97) [reply nnunetv2/training/dataloading/test_dataloader.py]: hmm ok will change. This part of the code is picked up as is from nnUNet.

File: nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py:1]: you mean the changes where I am printing time etc?

File: nnunetv2/training/dataloading/pytorch_nnunet_dataset.py (hunk 6 above)
Comment (ani0203-97) [reply nnunetv2/training/dataloading/pytorch_nnunet_dataset.py]: Moved this action to a separate function, but not deleted the sections for (1) and (2). I have added comments explaining what it does though.

Comment (abhijeetshenoi) [reply nnunetv2/training/dataloading/pytorch_nnunet_dataset.py]: Last question was answered by your response that explains `num_images_properties_loading_threshold`

File: nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py
Comment (abhijeetshenoi) [reply nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py:1]: No, the class `nnUNetTrainerPyTorchDataloader`

Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py:1]: i've currently structured it such that we still keep the old nnUNet trainer intact in case we need to use it. Can rename the new trainer to be more appropriate, but that change does not need to be removed no.
